Adopt dynamicDowncast<> in DragController

annevk opened the pull request:
#### bdd2db95c62291ce96f30b3ceb798403da21a53c
<pre>
Adopt dynamicDowncast&lt;&gt; in DragController
<a href="https://bugs.webkit.org/show_bug.cgi?id=268371">https://bugs.webkit.org/show_bug.cgi?id=268371</a>

Reviewed by Chris Dumez.

Also add some RefPtr.

* Source/WebCore/page/DragController.cpp:
(WebCore::elementUnderMouse):
(WebCore::DragController::tryDocumentDrag):
(WebCore::DragController::concludeEditDrag):
(WebCore::imageElementIsDraggable):
(WebCore::enclosingAttachmentElement):
(WebCore::DragController::draggableElement const):
(WebCore::getCachedImage):
(WebCore::DragController::startDrag):
(WebCore::DragController::insertDroppedImagePlaceholdersAtCaret):
* Source/WebCore/page/DragController.h:
(WebCore::DragController::protectedDocumentUnderMouse const):

Canonical link: <a href="https://commits.webkit.org/273832@main">https://commits.webkit.org/273832@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4189e0aed775254129180f10825bd929dd8994c6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36808 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15743 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39075 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39449 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32969 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38053 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12858 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31526 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37370 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13291 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32529 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11610 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11629 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32816 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40698 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33390 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33158 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37537 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11902 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9717 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35657 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13561 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8345 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12295 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12802 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented:
The failure looks real:
```
stderr:
ASSERTION FAILED: !m_inRemovedLastRefFunction
/Volumes/Data/worker/macOS-AppleSilicon-Sonoma-Debug-Build-EWS/build/Source/WebCore/dom/Node.h(799) : void WebCore::Node::ref() const
1   0x1390e3aa0 WTFCrash
2   0x282d10ee0 WebCore::AudioWorkletProcessor::buildJSArguments(JSC::VM&, JSC::JSGlobalObject&, JSC::MarkedVector<JSC::JSValue, 8ul, WTF::RecordOverflow>&, WTF::Vector<WTF::RefPtr<WebCore::AudioBus, WTF::RawPtrTraits<WebCore::AudioBus>, WTF::DefaultRefDerefTraits<WebCore::AudioBus>>, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc> const&, WTF::Vector<WTF::Ref<WebCore::AudioBus, WTF::RawPtrTraits<WebCore::AudioBus>>, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WTF::HashMap<WTF::String, std::__1::unique_ptr<WebCore::AudioArray<float>, std::__1::default_delete<WebCore::AudioArray<float>>>, WTF::DefaultHash<WTF::String>, WTF::HashTraits<WTF::String>, WTF::HashTraits<std::__1::unique_ptr<WebCore::AudioArray<float>, std::__1::default_delete<WebCore::AudioArray<float>>>>, WTF::MemoryCompactLookupOnlyRobinHoodHashTableTraits> const&)
3   0x283337990 WebCore::Node::ref() const
4   0x28018e38c WTF::DefaultRefDerefTraits<WebCore::Document>::refIfNotNull(WebCore::Document*)
5   0x282d408b4 WTF::RefPtr<WebCore::Document, WTF::RawPtrTraits<WebCore::Document>, WTF::DefaultRefDerefTraits<WebCore::Document>>::RefPtr(WebCore::Document*)
6   0x28018d3b0 WTF::RefPtr<WebCore::Document, WTF::RawPtrTraits<WebCore::Document>, WTF::DefaultRefDerefTraits<WebCore::Document>>::RefPtr(WebCore::Document*)
7   0x28446c600 WebCore::InspectorInstrumentation::instrumentingAgents(WebCore::ScriptExecutionContext&)
8   0x2833306fc WebCore::InspectorInstrumentation::instrumentingAgents(WebCore::ScriptExecutionContext*)
9   0x283fe0cb4 WebCore::InspectorInstrumentation::didChangeCanvasMemory(WebCore::CanvasRenderingContext&)
10  0x283fe0a74 WebCore::CanvasBase::setImageBuffer(WTF::RefPtr<WebCore::ImageBuffer, WTF::RawPtrTraits<WebCore::ImageBuffer>, WTF::DefaultRefDerefTraits<WebCore::ImageBuffer>>&&) const
11  0x28407913c WebCore::HTMLCanvasElement::createImageBuffer() const
```

annevk commented:
It is real, but it's https://github.com/WebKit/WebKit/pull/23495.

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -380,21 +380,23 @@ static bool isInShadowTreeOfEnabledColorInput(Node& node)
 
 // This can return null if an empty document is loaded.
-static Element* elementUnderMouse(Document* documentUnderMouse, const IntPoint& p)
+static RefPtr<Element> elementUnderMouse(Document& documentUnderMouse, const IntPoint& p)
```
I don't think this needs to return a smart pointer.

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -380,21 +380,23 @@ static bool isInShadowTreeOfEnabledColorInput(Node& node)
     // FIXME: Use parentElementInComposedTree here.
-    auto* element = is<Element>(*node) ? &downcast<Element>(*node) : node->parentElement();
-    auto* host = element->shadowHost();
+    RefPtr element = dynamicDowncast<Element>(*node);
```
So much ref'ing / deref'ing but not strictly necessary. Here we only call trivial getters on `element` so I don't think it needs to be a RefPtr.

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -380,21 +380,23 @@ static bool isInShadowTreeOfEnabledColorInput(Node& node)
+    RefPtr element = dynamicDowncast<Element>(*node);
+    if (!element)
+        element = node->parentElement();
+    RefPtr host = element->shadowHost();
```
We never deref host so I don't think it needs to be a RefPtr.

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -445,7 +447,7 @@ DragHandlingMethod DragController::tryDocumentDrag(LocalFrame& frame, const Drag
 
         IntPoint point = frameView->windowToContents(dragData.clientPosition());
-        Element* element = elementUnderMouse(m_documentUnderMouse.get(), point);
+        RefPtr element = elementUnderMouse(*m_documentUnderMouse, point);
```
This is still not safe, you need a RefPtr to the documentUnderMouse on the stack. So what I would do is:
1. Add a `RefPtr<Document> protectedDocumentUnderMouse() const` getter to the class
2. Call it here:  `RefPtr element = elementUnderMouse(*protectedDocumentUnderMouse(), point);`

annevk replied:
I have done as you suggested, but why not call protectedDocumentUnderMouse() at the top of this function and use the result instead of m_documentUnderMouse throughout?

cdumez replied:
If `m_documentUnderMouse` is used several times in the function and you're sure its value doesn't change, then storing it in a local is indeed better. You wouldn't call `protectedDocumentUnderMouse()` though. The `protectedFoo()` functions are there to *avoid* local variables. You would use:
```
RefPtr documentUnderMouse = m_documentUnderMouse;
```

annevk replied:
It does seem like it can change at least once so I'll leave that refactoring for someone more familiar with our drag & drop code. Thanks!

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -559,7 +561,7 @@ bool DragController::concludeEditDrag(const DragData& dragData)
 
     IntPoint point = m_documentUnderMouse->view()->windowToContents(dragData.clientPosition());
-    Element* element = elementUnderMouse(m_documentUnderMouse.get(), point);
+    RefPtr element = elementUnderMouse(*m_documentUnderMouse, point);
```
`*protectedDocumentUnderMouse()`

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -762,10 +764,11 @@ static bool imageElementIsDraggable(const HTMLImageElement& image, const LocalFr
         return false;
 
-    auto* cachedImage = downcast<RenderImage>(*renderer).cachedImage();
+    auto* cachedImage = renderImage->cachedImage();
```
Please use `CachedResourceHandle cachedImage = ...` to protect the cachedImage given the member function calls below.

cdumez commented on Source/WebCore/page/DragController.cpp:
```diff
@@ -1263,13 +1260,14 @@ bool DragController::startDrag(LocalFrame& src, const DragState& state, OptionSe
-    if (isColorControl && m_dragSourceAction.contains(DragSourceAction::Color)) {
-        auto& input = downcast<HTMLInputElement>(*state.source);
-        auto color = input.valueAsColor();
+    // FIXME: Should this use isEnabledColorInput instead?
```
Not a fan of adding questions in the code like this, unless something is obviously wrong. Here, it's not obvious to me there is an issue.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4189e0aed775254129180f10825bd929dd8994c6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/23502

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4189e0aed775254129180f10825bd929dd8994c6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/23502

annevk committed:
Adopt dynamicDowncast<> in DragController
https://bugs.webkit.org/show_bug.cgi?id=268371

Reviewed by Chris Dumez.

Also add some RefPtr.

* Source/WebCore/page/DragController.cpp:
(WebCore::elementUnderMouse):
(WebCore::DragController::tryDocumentDrag):
(WebCore::DragController::concludeEditDrag):
(WebCore::imageElementIsDraggable):
(WebCore::enclosingAttachmentElement):
(WebCore::DragController::draggableElement const):
(WebCore::getCachedImage):
(WebCore::DragController::startDrag):
(WebCore::DragController::insertDroppedImagePlaceholdersAtCaret):
* Source/WebCore/page/DragController.h:
(WebCore::DragController::protectedDocumentUnderMouse const):

Canonical link: https://commits.webkit.org/273832@main

webkit-commit-queue commented:
Committed 273832@main (bdd2db95c622): <https://commits.webkit.org/273832@main>

Reviewed commits have been landed. Closing PR #23502 and removing active labels.
